File: common/src/main/java/xyz/chunkstories/particle/ParticlesTypesStore.java
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.particle;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import xyz.chunkstories.api.content.Asset;
import xyz.chunkstories.api.content.Content;
import xyz.chunkstories.api.content.mods.ModsManager;
import xyz.chunkstories.api.exceptions.content.IllegalParticleDeclarationException;
import xyz.chunkstories.api.particles.ParticleTypeHandler;
import xyz.chunkstories.content.GameContentStore;

public class ParticlesTypesStore implements Content.ParticlesTypes {
	private final GameContentStore store;
	private final ModsManager modsManager;

	private static final Logger logger = LoggerFactory.getLogger("content.particles");

	public Logger logger() {
		return logger;
	}

	public ParticlesTypesStore(GameContentStore store) {
		this.store = store;
		this.modsManager = store.modsManager();

		// reload();
	}

	// private Map<Integer, ParticleTypeHandler> particleTypesById = new
	// HashMap<Integer, ParticleTypeHandler>();
	private Map<String, ParticleTypeHandler> particleTypesByName = new HashMap<String, ParticleTypeHandler>();

	public void reload() {
		// particleTypesById.clear();
		particleTypesByName.clear();

		//TODO review this system as well
		/*Iterator<Asset> i = modsManager.getAllAssetsByExtension("particles");
		while (i.hasNext()) {
			Asset f = i.next();
			loadParticlesFile(f);
		}*/
	}

	private void loadParticlesFile(Asset f) {
		if (f == null)
			return;
		throw new UnsupportedOperationException("TODO");
		/*try (BufferedReader reader = new BufferedReader(f.reader());) {
			String line = "";
			while ((line = reader.readLine()) != null) {
				if (line.startsWith("#")) {
					// It's a comment, ignore.
				} else {
					if (line.startsWith("end")) {
						logger().warn("Syntax error in file : " + f + " : ");
						continue;
					}
					String splitted[] = line.split(" ");
					if (splitted.length == 2 && splitted[0].startsWith("particle")) {
						// int id = Integer.parseInt(splitted[2]);
						String particleName = splitted[1];

						try {
							ParticleTypeDefinitionImplementation type = new ParticleTypeDefinitionImplementation(this,
									particleName, reader);
							ParticleTypeHandler handler = type.handler();

							particleTypesByName.put(particleName, handler);
							// particleTypesById.put(id, handler);

						} catch (IllegalParticleDeclarationException e) {
							this.store.getContext().logger()
									.error("Could not load particle type " + particleName + " : \n" + e.getMessage());
							e.printStackTrace();
						}
					}
				}
			}
		} catch (IOException e) {
			logger().warn(e.getMessage());
		}*/
	}

	@Override
	public ParticleTypeHandler getParticleType(String string) {
		return particleTypesByName.get(string);
	}

	/*
	 * @Override public ParticleTypeHandler getParticleTypeHandlerById(int id) {
	 * return particleTypesById.getVoxelComponent(id); }
	 */

	@Override
	public Iterator<ParticleTypeHandler> all() {
		return this.particleTypesByName.values().iterator();
	}

	@Override
	public Content parent() {
		return store;
	}

}

File: common/src/main/java/xyz/chunkstories/particle/ParticlesTypesStore.kt
//
// This file is a part of the Chunk Stories Implementation codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.particle

import com.google.gson.Gson
import com.google.gson.internal.LinkedTreeMap
import org.hjson.JsonValue
import org.slf4j.LoggerFactory
import xyz.chunkstories.api.content.Asset
import xyz.chunkstories.api.content.Content
import xyz.chunkstories.api.content.json.asDict
import xyz.chunkstories.api.content.json.asString
import xyz.chunkstories.api.particles.ParticleType
import xyz.chunkstories.api.particles.ParticleTypeDefinition
import xyz.chunkstories.content.GameContentStore
import xyz.chunkstories.content.eat
import xyz.chunkstories.content.extractProperties
import java.util.*

class ParticlesTypesStore(override val parent: GameContentStore) : Content.ParticlesTypes {
    override val logger = LoggerFactory.getLogger("content.particles")

    private val particleTypesByName = HashMap<String, ParticleType<*>>()

    fun reload() {
        particleTypesByName.clear()

        val gson = Gson()

        fun readDefinitions(a: Asset) {
            logger.debug("Reading particle definitions in : $a")

            val json = JsonValue.readHjson(a.reader()).eat().asDict ?: throw Exception("This json isn't a dict")
            val dict = json["particles"].asDict ?: throw Exception("This json doesn't contain an 'particles' dict")

            for (element in dict.elements) {
                val name = element.key
                val properties = element.value.asDict ?: throw Exception("Definitions have to be dicts")

                try {
                    val particleTypeDefinition = ParticleTypeDefinition(this, name, properties)

                    val className = particleTypeDefinition["class"].asString ?: throw Exception("no 'class' property set")
                    val klass = parent.modsManager.getClassByName(className) ?: throw Exception("Class $className not found")
                    val constructor = klass.getConstructor(ParticleTypeDefinition::class.java)
                            ?: throw Exception("$className doesn't have the right constructor")

                    val particleType = constructor.newInstance(particleTypeDefinition) as ParticleType<*>
                    particleTypesByName.put(name, particleType)

                    logger.debug("Loaded particle type $particleType")
                } catch (e: Exception) {
                    logger.error("Can't load particle $name: ${e.message}")
                }
            }
        }

        for (asset in parent.modsManager.allAssets.filter { it.name.startsWith("particles/") && it.name.endsWith(".hjson") }) {
            readDefinitions(asset)
        }
    }

    private fun loadParticlesFile(f: Asset?) {
        if (f == null)
            return
        throw UnsupportedOperationException("TODO")
    }

    override fun <T : ParticleType.Particle> getParticleType(string: String): ParticleType<T>? {
        return particleTypesByName[string] as? ParticleType<T>
    }

    override val all: Collection<ParticleType<*>>
        get() = particleTypesByName.values
}
